core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java

```diff
@@ -114,14 +115,29 @@ public void close() throws IOException {
 
   @Override
   public StreamObserver<WriteRequest> writeBlock(StreamObserver<WriteResponse> responseObserver) {
-    return mStreamingAsyncStub.writeBlock(responseObserver);
+    if (responseObserver instanceof DataMessageMarshallerProvider) {
+      DataMessageMarshaller<WriteRequest> marshaller =
+          ((DataMessageMarshallerProvider<WriteRequest, WriteResponse>) responseObserver)
+              .getRequestMarshaller();
+      Preconditions.checkNotNull(marshaller);
+      return mStreamingAsyncStub
+          .withOption(GrpcSerializationUtils.OVERRIDDEN_METHOD_DESCRIPTOR,
+              BlockWorkerGrpc.getWriteBlockMethod().toBuilder()
+                  .setRequestMarshaller(marshaller)
+                  .build())
+          .writeBlock(responseObserver);
+    } else {
+      return mStreamingAsyncStub.writeBlock(responseObserver);
+    }
   }
 
   @Override
   public StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> responseObserver) {
     if (responseObserver instanceof DataMessageMarshallerProvider) {
       DataMessageMarshaller<ReadResponse> marshaller =
-          ((DataMessageMarshallerProvider<ReadResponse>) responseObserver).getMarshaller();
+          ((DataMessageMarshallerProvider<ReadRequest, ReadResponse>) responseObserver)
+              .getResponseMarshaller();
+      Preconditions.checkNotNull(marshaller);
```

Comment:
Similar to above, if user does not want to use zero copy for reading blocks, they should not pass  `DataMessageMarshallerProvider` instead of giving a null response marshaller.